src/util/invariant.h

```diff
@@ -51,14 +53,63 @@ Author: Martin Brain, martin.brain@diffblue.com
 ** CPROVER_INVARIANT_* macros.
 */
 
+/// A logic error, augmented with a distinguished field to hold a backtrace.
 class invariant_failedt: public std::logic_error
 {
 public:
-  explicit invariant_failedt(const std::string& what) : logic_error(what) {}
-  explicit invariant_failedt(const char* what) : logic_error(what) {}
+  invariant_failedt(
+    const std::string& what):
+  logic_error(what),
+  backtrace()
+  {
+  }
+
+  /// Stores the backtrace set by `invariant_violated`.
+  std::string backtrace;
+
+  /// Set the backtrace stored against this error report.
+  /// This is overwritten by `invariant_violated`, which is called
+  /// from the `INVARIANT` family of macros.
+  void set_backtrace(const std::string &_backtrace)
+  {
+    backtrace=_backtrace;
+  }
+
+  /// Return a human-readable description of this error.
+  /// For this base class that's just the description string
+  /// supplied to the constructor; subclasses may override this to
+  /// provide a description in a different way, e.g. from a more
+  /// structured error description. `std::logic_error::what` is
+  /// overridden to use this and the backtrace to describe the
+  /// error.
+  virtual std::string pretty_print() const
+  {
+    return std::logic_error::what();
+  }
+
+  /// Temporary storage for `what`'s return value
+  mutable std::string what_str;
```

Comment:
The string needs to live beyond the lifespan of `what()`. It's a bad interface; the docs for `std::logic_error` admit as much, but I'd rather produce this on demand than inevitably have some derived class mutator forget to update it.